DevDocs\Theme: Add ThemesList example app-component

afsheenirani opened the pull request:
## Background

- Add the `ThemesList` application component example to the [devdocs\design\app-components](https://wpcalypso.wordpress.com/devdocs/app-components) page.
- Also, make the list of application components in alphabetical in order to have a better visual experience when applying search filtering at the top of the page.

## Testing
1. Load the [app-components](http://calypso.localhost:3000/devdocs/app-components) page.
2. Scroll down to ensure you see the `Themes List` example.

![image](https://cloud.githubusercontent.com/assets/14303849/13371509/ef8ae734-dcf5-11e5-8d94-f58e55976239.png)
3. Search for `Themes` in the search box at the top of the page to ensure filtering works.
4. Click on the `Themes List` link to make sure it takes you to a page with ONLY the `Themes List` component i.e., [app-components/themes-list](http://calypso.localhost:3000/devdocs/app-components/themes-list).
5. Hover over each of the three Themes in the `ThemesList` component. Make sure this shows you individual Action labels for each theme e.g. `CLICK ACTION THEME 1` or `CLICK ACTION THEME 2`.
6. Click on each theme and view the developer console logs to make sure it prints individual messages for each theme e.g., `Theme twentyfourteen at 0 clicked`. This tells you the theme id and index of that theme in the `ThemesList` that was clicked.
7. Click on the More Buttons `...` link on any of the themes to make sure they work and when you click one of the links that pop-up you should see a message in the Developer console as `Menu Item 1 selected`.

## Reviewers
@designsimply, @nb, @folletto, @ockham 

## Issues
Closes #2119

mcsf commented on client/devdocs/design/app-components.jsx:
```diff
@@ -1,25 +1,29 @@
+import HeaderCake from 'components/header-cake';
+import SearchCard from 'components/search-card';
+
+// Example App Components
```
The new import order can be kept, but, for consistency, this comment and its accompanying newline should be removed.

afsheenirani replied:
Will do.

mcsf commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,77 @@
+	displayName: 'ThemesList',
+
+	render: function() {
+		const demoThemes = [ {
```
Like `onScreenshotClickFn`, `demoThemes` can benefit from being placed up in the module scope.

afsheenirani replied:
Makes sense, done.

mcsf commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,77 @@
+
+} );
+
+module.exports = ThemesListExample;
```
We should stick to ES6-style imports and exports.

afsheenirani replied:
Agreed, done.

mcsf commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,77 @@
+		);
+	}
+
+} );
```
Throughout this file, let's prefer ES6 arrow functions for the anonymous functions (`theme => theme.actionLabel`) and method shorthand for component methods (`render() { … }`).

mcsf replied:
Also, all the functions provided to `ThemesList` are unnecessarily re-instantiated on each render, so let's make them method of `ThemesListExample` instead and pass them to `ThemesList`.

afsheenirani replied:
Incorporated this feedback as well.

mcsf commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,77 @@
+					<a href="/devdocs/app-components/themes-list">Themes List</a>
+				</h2>
+				<ThemesList
+				themes={ demoThemes }
```
Starting with this line, the props passed to `ThemesList` should be indented one extra level.

afsheenirani replied:
Done.

mcsf commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,77 @@
+};
+
+const ThemesListExample = React.createClass( {
+	displayName: 'ThemesList',
```
This seems like improper use of `displayName`, as our component here just wraps around the real `ThemesList`, so no need to disguise it. And since the previous line immediately binds the component to `const ThemesListExample`, React might be smart enough not to need `displayName` at all.

So I'd either remove the property or rename its value to `ThemesListExample`.

afsheenirani replied:
So I had originally changed this to `ThemesListExample`, but it does not work when you click the **Themes List** heading to view only that one app component. Just recalled and realized that my latest commit still has that issue. Let me look at some other examples to see how that has been handled.

afsheenirani replied:
So the issue here is that the `displayName` is used in determining what components are to be displayed.

1. When you click the header link for any one component it uses the name of the component that _**you**_ supply as part of the `<a href="/devdocs/app-components/`**themes-list**`">` link provided. This is then passed down as the `this.props.component` to the `app-components.jsx` as seen in `client/devdocs/controller.js`. This is the _**selected**_ component.
2. Now in `client\devdocs\design\app-components.jsx --> Collection.shouldWeHide()`, the Collection determines which children components need to be displayed based on whether the `displayName` of any of the children components matches the name of the above selected component.
3. In our scenario, what we end up with is: `themeslist` does not match `themeslistexample` and this results in the component being hidden.

To solve this, we have two options:

1. Keep the `<a href>` header link in line with the path of where the component is defined e.g., `<a href="/devdocs/app-components/themes-list">` and then set the displayName of the Example class to `ThemesList`.

 **Note:** This seems to be the methodology that is being followed (as seen in other component examples).
2. Keep the `displayName` as `ThemesListExample` and update the `<a href>` path to `themes-list-example`.

 **Note:** I find this method cleaner/better, but it will be different from the other examples i've seen previously.

I'll submit my change following Option 2, we can further discuss if there are any other better options.

mcsf commented:
Thanks for the PR! The changes look good overall, though I suggested some improvements.

afsheenirani commented:
@mcsf - Firstly, thank you very much for that detailed review. Really appreciate all the good feedback...especially around using ES6 standards/conventions. I'll certainly incorporate that in all my changes going forward.

I've incorporated all changes and have also gone into some detail in explaining the issue around why I had set the `displayName` to `ThemesList` and not `ThemesListExample`. However, I've since figured out a way to set the `displayName` to `ThemesListExample` without having any side effects and have made the necessary changes. 

Let me know what you think...

afsheenirani commented:
@mcsf - Hi Miguel - Do you have any further comments/recommendations based on my latest updates?

mcsf commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,76 @@
+		}
+	},
+
+	onScreenshotClick: ( theme, index ) => theme.onScreenshotClick( theme, index ),
```
This, along with `getActionLabel`, should read:

```js
onScreenshotClick( theme, index ) {
  theme.onScreenshotClick( theme, index );
}
```

That is, use class method shorthand. However, now looking at the implementation, this is pretty redundant. I'd either bind each theme's `onScreenshotClick` to the theme itself upon instantiation of `demoThemes`, _or_ I'd skip copying `onScreenshotClickFn` into each theme object and work with `onScreenshotClickFn` directly in this method here.

mcsf commented on client/devdocs/design/app-components.jsx:
```diff
@@ -1,25 +1,27 @@
+import Theme from 'components/theme/docs/example';
+import ThemesListExample from 'components/themes-list/docs/example';
+
+var Collection,
```
While we're here, this can be a `let` statement. :)

mcsf commented:
Hey Afsheen, thanks for the changes and for detailing your process. The PR makes sense, and I've pointed out some additional stuff. However, I only managed to perform a drive-by review and I am currently offline, so unable to test. 

Furthermore, a team switch has just made Themes a lesser priority for me, so it makes more sense to e.g. ping @ockham. :)

afsheenirani commented:
Good eye @mcsf! Clearly I was already instantiating the individual Themes with an onScreenShotClick func and then passing them to ThemesList. While I was also passing an onScreenShotClick func to ThemesList as well.

I guess that we should only be leveraging the onScreenShotClick to instantiate the ThemesList component...which will then use it to setup the onScreenShotClick for the individual themes in the themes list.

I've made the necessary changes to remove this redundancy and then re-tested it.

@ockham - Would you mind reviewing my changes?

afsheenirani commented:
Also cc'ing @apeatling @seear @nb for review based on original contributions to ThemesList.

mcsf commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,71 @@
+
+	onScreenshotClickFn( theme, index ) {
+		console.log( `Theme ${ theme.id } at ${ index } clicked` );
+	},
```
Final touch would be to avoid `onFoo` method names. It's acceptable to use `onFoo` as a prop name when a component expects a function to be passed to it (like `ThemesList`), but when we implement such a function we prefer a descriptive name. So a candidate here would be `logThemeClick`, for instance.

afsheenirani replied:
Feedback duly noted. I did find it confusing to deal with `onScreenShotClick` of `ThemesList`, `onScreenShotClick` of `Theme` and then `onScreenShotClickFn` as the func to `ThemesList`. 

I will definitely incorporate this suggestion to avoid confusion around func prop naming.

mcsf commented:
Hey, Afsheen. This looks nice. FYI: @ockham has been in a team meetup, along with me, so he might take a while to get back to you.

afsheenirani commented:
Thanks @mcsf! Just re-based, squashed commits and re-tested. Will await @ockham's review...cheers!

ockham commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,71 @@
+			action1: {
+				label: 'Menu Item 1',
+				action: function() {
+					console.log( 'Menu Item 1 selected' );
```
It'd be nice if this message also showed which theme this was selected for. (`'Menu item 1 for theme <themeslug> selected'`)

ockham commented on client/components/themes-list/docs/example.jsx:
```diff
@@ -0,0 +1,71 @@
+				</h2>
+				<ThemesList
+					themes={ demoThemes }
+					fetchNextPage={ noop }
```
Might make sense to add this as a default prop to the `ThemesList` component proper.

afsheenirani replied:
@ockham - fetchNextPage is defined as a property that is isRequired on ThemesList. However, it also has a default property setup to be an empty function.

I'll need to check if we can safely remove the fetchNextPage property during ThemesList instantiation. I am not sure whether React's property validation runs before setting up the default properties or after. Will research and also test for what the expected behavior is and then action accordingly.

afsheenirani replied:
@ockham - Based on my testing, and the information provided [here](https://facebook.github.io/react/docs/reusable-components.html), I have removed the fetchNextPage property during the instantiation of the `ThemesList` component.

ockham commented:
Left two minor notes. Other than that, looks pretty good.

afsheenirani commented:
@ockham - I've made the suggested edits and resubmitted. Thanks for the review!

ockham commented:
Looks good. One last thing -- please squash those two commits into one and change the commit message (specifically removing those `# This is a combination of 2 commits` lines).

afsheenirani commented:
@ockham - Done. Squashed commits and added appropriate commit comments.

Btw, for my future reference, is it always the case that the final release i.e. merge is expected to be a single commit (if so, I can just pro-actively always squash my commits).

mcsf commented:
@afsheenirani: not necessarily. Ideally, there should be a reasonably-sized set of atomic and self-descriptive commits, maybe for each feature, or each improvement, etc. Now, as for how to define atomic and what the desired granularity of the commits is, one should simply exercise judgement. There will be disagreements, and that's fine.

ockham commented:
> Btw, for my future reference, is it always the case that the final release i.e. merge is expected to be a single commit (if so, I can just pro-actively always squash my commits).

What @mcsf said :-)

And in this particular case, the second commit was just an addendum to the first, since it was addressing my feedback -- but it wasn't implementing anything really independent of the first. And a fix like that is best merged into the commit that implements the actual change.

(Re-ordering application components, OTOH, could be seen as a change that is independent of implementing a new example component, and could thus warrant a separate commit. This is mostly relevant if a commit breaks something -- ideally we'd then want to revert the offending commit as fine-grainedly as possible. But at this point, I wouldn't go through the pain of splitting that commit along those lines.)

Looks good. :+1: I'll merge this.

afsheenirani commented:
Thanks @mcsf and @ockham for the feedback on the reviews and the commit workflow...totally makes sense...will employ similar logic for future commits/reviews.
